aesd-char-driver/main.c

```diff
@@ -30,186 +33,297 @@ struct aesd_dev aesd_device;
+		retval = copy_from_user((void *) dev->entry.buffptr + dev->entry.size, buf, count);
+
+		retval = count - retval;
+		dev->entry.size += retval;
+		PDEBUG("copied %ld bytes from userspace to kernel space, total size %ld\n", \
+					retval, dev->entry.size);
+
+		if (dev->entry.buffptr[(dev->entry.size - 1)] == '\n') {
+			rtnptr = aesd_circular_buffer_add_entry(&dev->cb, &dev->entry);
+			if (rtnptr != NULL)
+				kfree(rtnptr);
+
+			dev->entry.buffptr = NULL;
+			dev->entry.size = 0;
+		}
+	}
+
+	mutex_unlock(&dev->lock);
+
+	return retval;
+}
+
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+	long retval = 0;
+	int index = 0;
+	struct aesd_buffer_entry *entry = NULL;
+	struct aesd_dev *dev = NULL;
+
+	PDEBUG("adjust_file_offset\n");
+
+	if (filp == NULL) {
+		PDEBUG("invalid arguments\n");
+		return -EINVAL;
+	}
+
+	dev = filp->private_data;
+
+	if (mutex_lock_interruptible(&dev->lock) != 0) {
+		PDEBUG("failed to acquire mutex\n");
+		return -ERESTARTSYS;
+	}
+
+	AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->cb, index) {
+		/* index will give the total number of commands in circular buffer */
+	}
+
+	if (write_cmd > index ||
+		write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED ||
+		write_cmd_offset >= dev->cb.entry[write_cmd].size) {
+		retval = -EINVAL;
+	} else {
+		for (index = 0; index < write_cmd; index++)
+			filp->f_pos += dev->cb.entry[index].size;
+
+		filp->f_pos += write_cmd_offset;
+	}
+
+	mutex_unlock(&dev->lock);
+
+	return retval;
+}
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+	long retval = 0;
+	struct aesd_seekto seekto;
+
+	PDEBUG("ioctl\n");
+
+	if (filp == NULL) {
+		PDEBUG("invalid arguments\n");
+		return -EINVAL;
+	}
+
+   	/*
+	 * extract the type and number bitfields, and don't decode
+	 * wrong cmds: return ENOTTY (inappropriate ioctl) before access_ok()
+	 */
+	if (_IOC_TYPE(cmd) != AESD_IOC_MAGIC) return -ENOTTY;
+	if (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR) return -ENOTTY;
+
+	switch (cmd) {
+	case AESDCHAR_IOCSEEKTO:
+		if (copy_from_user(&seekto, (const void __user *)arg, sizeof(seekto)) != 0)
+			retval = -EFAULT;
+		else
+			retval = aesd_adjust_file_offset(filp, seekto.write_cmd, seekto.write_cmd_offset);
+		break;
+
+	default:
+		retval = -ENOTTY;
+		break;
+	}
+
+	return retval;
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+	loff_t newpos;
+	loff_t size = 0;
+	int index = 0;
+	struct aesd_dev *dev = NULL;
+	struct aesd_buffer_entry *entry = NULL;
+
+	PDEBUG("llseek\n");
+
+	if (filp == NULL) {
+		PDEBUG("invalid arguments\n");
+		return -EINVAL;
+	}
+
+	dev = filp->private_data;
+
+	if (mutex_lock_interruptible(&dev->lock) != 0) {
+		PDEBUG("failed to acquire mutex\n");
+		return -ERESTARTSYS;
+	}
+
+	/* calculate size */
+	AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->cb, index)
+		size += entry->size;
+
+	newpos = fixed_size_llseek(filp, off, whence, size);
```

Comment:
@hgajapathy does `fixed_size_llseek` need to be in your locking interval?  What is this lock locking?

Reply:
No fixed_size_llseek  doesn't need to be be locked. Thank you for the comment. I will fix it.